File: 14 Longest Common Prefix/14longest-common-prefix.py
class Solution:
    def longestCommonPrefix(self, strs: List[str]) -> str:
        res = ""

        for i in range(len(strs[0])):
            for word in strs:
                if i >= len(word) or word[i] != strs[0][i]:
                    return res
            res += word[i]
        return res

# Time: O(nl). n: len(strs), l: shortest len (word)
# Space: O(l). 

File: 14 Longest Common Prefix/14longest-common-prefix.cpp
class Solution {
public:
    string longestCommonPrefix(vector<string>& strs) {
        
        for (int i = 0; i < strs[0].size(); i++) {
            for (const string& str : strs) {
                if (i >= str.size() or strs[0][i] != str[i]) {
                    return str.substr(0, i);
                }
            }
        }
        return strs[0];
    }
};

// Time: O(n*w). w: len(shortest word)
// Space: O(w)